packages/d3fc-webgl/src/series/glLine.js

```diff
@@ -0,0 +1,180 @@
+    const yPrevValueAttrib = 'aPrevYValue';
+    const cornerValueAttrib = 'aCorner';
+    const widthUniform = 'uWidth';
+    const screenUniform = 'uScreen';
+
+    const draw = (numElements, segments = []) => {
+        if (!segments.length) {
+            segments = [{
+                numElements: numElements,
+                start: 0,
+                bufferSize: numElements
+            }];
+        }
+        // we are resetting the shader each draw here, to avoid issues with decorate
+        // we'll eventually need a way to change the symbol type here
+        const shaderBuilder = lineShader();
+        program.vertexShader(shaderBuilder.vertex())
+               .fragmentShader(shaderBuilder.fragment())
+               .mode(drawModes.TRIANGLE_STRIP);
+
+        // apply the scale to our current, next and previous coordinates
+        xScale.coordinate(0);
+        xScale(program);
+        yScale.coordinate(1);
+        yScale(program);
+        xScale.scaleComponent(program, 'next');
+        yScale.scaleComponent(program, 'next');
+        xScale.scaleComponent(program, 'prev');
+        yScale.scaleComponent(program, 'prev');
+
+        // from here we are dealing with pre-scaled vertex positions
+        program.vertexShader()
+            .appendBody(`if (all(equal(gl_Position.xy, prev.xy))) {
+                prev.xy = gl_Position.xy + normalize(gl_Position.xy - next.xy);
+            }`)
+            .appendBody(`if (all(equal(gl_Position.xy, next.xy))) {
+                next.xy = gl_Position.xy + normalize(gl_Position.xy - prev.xy);
+            }`)
+            .appendBody(`vec2 A = normalize(normalize(gl_Position.xy - prev.xy) * uScreen);`)
+            .appendBody(`vec2 B = normalize(normalize(next.xy - gl_Position.xy) * uScreen);`)
+            .appendBody(`vec2 tangent = normalize(A + B);`)
+            .appendBody(`vec2 miter = vec2(-tangent.y, tangent.x);`)
+            .appendBody(`vec2 normalA = vec2(-A.y, A.x);`)
+            .appendBody(`float miterLength = 1.0 / dot(miter, normalA);`)
+            .appendBody(`vec2 point = normalize(A - B);`)
+            .appendBody(`if (miterLength > 10.0 && sign(aCorner.x * dot(miter, point)) > 0.0) {
+                gl_Position.xy = gl_Position.xy - (sign(dot(normalA, miter)) * aCorner.x * aCorner.y * uWidth * normalA) / uScreen.xy;
+            } else {
+                gl_Position.xy = gl_Position.xy + (aCorner.x * miter * uWidth * miterLength) / uScreen.xy;
+            }`);
+
+        program.buffers().uniform(screenUniform, uniformBuilder([
+            program.context().canvas.width,
+            program.context().canvas.height
+        ]));
+
+        width()(program);
+
+        decorate(program);
+
+        segments.forEach(segment => {
+            // we're sending every vertex to the shader four times, hence * 4
+            program(segment.numElements * 4, segment.start * 4, segment.bufferSize * 4);
+        });
+    };
+
+    draw.xValues = (...args) => {
```

Comment:
This line function is understandably quite complicated, do you have a link to any documentation that you followed when you were implementing it that you could add in as a comment to help us to understand it? :)